text/0000-async_await.md

```diff
@@ -0,0 +1,571 @@
+shipping - async/await syntax (a concept available widely in many languages
+which interacts well with our existing async IO libraries) is the most logical
+thing to implement at this stage in Rust's evolution.
+
+# Prior art
+[prior-art]: #prior-art
+
+There is a lot of precedence from other languages for async/await syntax as a
+way of handling asynchronous operation - notable examples include C#,
+JavaScript, and Python.
+
+There are three paradigms for asynchronous programming which are dominant
+today:
+
+- Async and await notation.
+- An implicit concurrent runtime, often called "greenthreading," such as
+  communicating sequential processes (e.g. Go) or an actor model (e.g. Erlang).
+- Monadic transformations on lazily evaluated code, such as do notation (e.g.
+  Haskell).
+
+Async/await is the most compelling model for Rust because it interacts
+favorably with ownership and borrowing (unlike systems based on monads) and it
+enables us to have an entirely library-based asynchronicity model (unlike
+greenthreading).
+
+One way in which our handling of async/await differs from most other statically
+typed languages (such as C#) is that we have chosen to show the "inner" return
+type, rather than the outer return type. As discussed in the alternatives
+section, Rust's specific context (lifetime elision, the lack of a need for
+return type polymorphism here) make this deviation well-motivated.
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+This section contains design extensions which have been postponed & not
+included in this initial RFC.
+
+## Final syntax for the `await` expression
+
+Though this RFC proposes that `await` be a built-in macro, we'd prefer that
+some day it be a normal control flow construct. The unresolved question about
+this is how to handle its precedence & whether or not to require delimiters of
+some kind.
+
+In particular, `await` has an interesting interaction with `?`. It is very
+common to have a future which will evaluate to a `Result`, which the user will
+then want to apply `?` to. This implies that await should have a tighter
+precedence than `?`, so that the pattern will work how users wish it to.
+However, because it introduces a space, it doesn't look like this is the
+precedence you would get:
+
+```
+await future?
+```
+
+There are a couple of possible solutions:
+
+1. Require delimiters of some kind, maybe braces or parens or either, so that
+   it will look more like how you expect - `await { future }?` - this is rather
+   noisy.
+2. Define the precedence as the obvious, if inconvient precedence, requiring
+   users to write `(await future)?` - this seems very surprising for users.
+3. Define the precedence as the inconvient precedence - this seems equally
+   surprising as the other precedence.
+4. Introduce a special syntax to handle the multiple applications, such as
+   `await? future` - this seems very unusual in its own way.
```

Comment:
What about Bar(Foo() await) ? :p

Reply:
Could you elaborate on why it should be postfix?

I think `await Foo()` reads better for an English (and at least Scandinavian) reader since you are (a)waiting for `Foo()` to complete.

Reply:
It solves the precedence confusion around `?` and makes chaining easier. Basically the same arguments for going from `try!` to `?`.

Reply:
@Centril I agree it looks better for simple cases, but once you have multiple or need to mix it with methods and fields, it gets worse.  I think `x.foo().await.bar().await.qaz` (with highlighting) will look way better than `(await (await x.foo()).bar()).qaz` -- fewer hard-to-track-down parens and maintains the reading order.

(Prefix would probably be better in Lisp or Haskell, but postfix seems generally better in rust.  I sometimes even wish we had _no_ prefix operators, so there'd never be a `!(foo?)` vs `(!foo)?` confusion...)

Reply:
Like with the `?` instead of `try!`, it puts the actual thing that we are doing at the first place, and boilerplaty technicalities (yield point or not, error handling) at the end.

Reply:
I like the idea of a postfix await operator, but I'm not sure what the best syntax for that would be.
`f.await` looks too much like accessing a field `f.await()` looks too much like a method call `f await` is just as awkward to use as `await f`.

Reply:
Or it can be just some another special character, like `future@?` or `future~?`.